components/scream/src/control/surface_coupling.cpp

```diff
@@ -379,60 +369,49 @@ get_col_info(const std::shared_ptr<const FieldHeader>& fh,
   col_offset = vecComp;
   if (layout3d) {
     if (lt==LayoutType::Vector3D) {
-      col_offset *= fh->get_alloc_properties().get_last_extent();;
+      col_offset *= fh->get_alloc_properties().get_last_extent();
     }
-    col_offset += m_num_levs-1;
+    col_offset += dims.back()-1;
   }
 
   // If rank>1, there always is some stride
   col_stride = 1;
   if (layout.rank()>1) {
-    col_stride = fh->get_alloc_properties().get_last_extent();;
+    col_stride = fh->get_alloc_properties().get_last_extent();
     if (layout.rank()>2) {
       col_stride *= dims[1];
     }
   }
 
+  // If this field has a parent, then the underlying data includes all entires
+  // in the larger field. We then must treat the child field as having col_stride
+  // and col_offset of a vector field with component as its subview_idx.
+  std::shared_ptr<const FieldHeader> parent = fh->get_parent().lock();
+  if (parent != nullptr) {
 
-  // Recursively go through parent field, adding offset depending
-  // on the location of the slice.
-  // Note: as of now, I don't expect there to be "a grandparent" for this field,
-  //       so we could avoid the while loop. However, the logic is so simple, that
-  //       I may as well add it, and be safe down the road if it happens.
-
-  // Product of dimensions of field, without counting space dims, that is,
-  // the product of dimensions of the "analytical" field.
-  int dims_prod = 1;
-    if (vector) {
-      dims_prod = dims[1];
-    }
+    EKAT_REQUIRE_MSG(parent->get_parent().lock() == nullptr,
+                     "Error! Currently support isn't added for fields with grandparents.\n");
 
-  std::shared_ptr<const FieldHeader> me = fh;
-  std::shared_ptr<const FieldHeader> p  = me->get_parent().lock();
-  while (p!=nullptr) {
+    const auto parent_lt = get_layout_type(parent->get_identifier().get_layout().tags());
 
-    // TODO: Currently this is not used as we require fields to not have parents in
-    //       register_import/export() functions. It is possible, but the logic here
-    //       would need to be tested. We will end up erroring out before making it here.
+    // If the field has a parent field, it should be the case that the
+    // parent field is a vectored field. Currently we conly expect
+    EKAT_REQUIRE_MSG(parent_lt==LayoutType::Vector3D,
+                     "Error! SurfaceCoupling expects all subfields to have parents"
+                     " with LayoutType::Vector3D.\n");
 
-    const auto& idx = me->get_alloc_properties().get_subview_idx();
+    const auto& idx = fh->get_alloc_properties().get_subview_idx();
 
     // Recall: idx = (idim,k) = (dimension where slice happened, index along said dimension).
     // Field class only allows idim=0,1. But we should never be in the case of idim=0, here.
     // If we have idim=0, it means that the parent field did not have CL has tag[0]...
```

Comment:
While at it, would you mind fixing this comment? I think the last sentence was supposed to be "... did not have COL as tag[0]" (couple of typos).